src/OrleansTelemetryConsumers.Counters/OrleansPerfCounterTelemetryConsumer.cs

```diff
@@ -0,0 +1,235 @@
+        {
+            isInstalling = true;
+            if (!isInitialized.Value)
+            {
+                var msg = "Unable to install Windows Performance counters";
+                logger.Warn(ErrorCode.PerfCounterNotFound, msg);
+                throw new InvalidOperationException(msg);
+            }
+
+            var collection = new CounterCreationDataCollection();
+            collection.AddRange(GetCounterCreationData());
+
+            PerformanceCounterCategory.Create(
+                CATEGORY_NAME,
+                CATEGORY_DESCRIPTION,
+                PerformanceCounterCategoryType.SingleInstance,
+                collection);
+        }
+
+        /// <summary>
+        /// Delete any existing perf counters registered with Windows
+        /// </summary>
+        /// <remarks>Note: Program needs to be running as Administrator to be able to delete Windows perf counters.</remarks>
+        public void DeleteCounters()
+        {
+            PerformanceCounterCategory.Delete(CATEGORY_NAME);
+        }
+
+        private PerfCounterConfigData GetCounter(string counterName)
+        {
+            return perfCounterData.Where(pcd => GetPerfCounterName(pcd) == counterName).SingleOrDefault();
+        }
+
+        #endregion
+
+        #region IMetricTelemetryConsumer Methods
+
+        public void IncrementMetric(string name) => WriteMetric(name, UpdateMode.Increment);
+
+        public void IncrementMetric(string name, double value) => WriteMetric(name, UpdateMode.Increment, value);
+
+        public void TrackMetric(string name, double value, IDictionary<string, string> properties = null) => WriteMetric(name, UpdateMode.Set, value);
+
+        public void TrackMetric(string name, TimeSpan value, IDictionary<string, string> properties = null) => WriteMetric(name, UpdateMode.Set, value.Ticks);
+
+        public void DecrementMetric(string name) => WriteMetric(name, UpdateMode.Decrement);
+
+        public void DecrementMetric(string name, double value) => WriteMetric(name, UpdateMode.Decrement, value);
+
+        public void Flush() { }
+
+        public void Close() { }
+
+        private void WriteMetric(string name, UpdateMode mode = UpdateMode.Increment, double? value = null)
+        {
+            if (!isInitialized.Value)
+                return;
+
+            PerfCounterConfigData cd = GetCounter(name);
+            if (cd == null)
+                return;
+
+            StatisticName statsName = cd.Name;
+            string perfCounterName = GetPerfCounterName(cd);
+
+            try
+            {
+                
+                if (logger.IsVerbose3) logger.Verbose3(ErrorCode.PerfCounterWriting, "Writing perf counter {0}", perfCounterName);
+
+                switch (mode)
+                {
+                    case UpdateMode.Increment:
+                        if (value.HasValue)
+                        {
+                            cd.PerfCounter.IncrementBy((long)value.Value);
```

Comment:
I think original code checked if cd  was null, then if cd.PerfCounter was null.  If cd.PerfCounter was null it would create it, which we don't want to do, but if it is null, we don't want to use it.